### DISCUSSION TITLE
Shape the future of snoots

### DISCUSSION THREAD

**Discussion by thislooksfun** [thislooksfun/snoots#4] [General]
Thank you for using snoots! It's still under *heavy* development, so there's a lot of work left to be done. This involves making some hard decisions about how to expose the api to you, and I'd love you all to weigh in with your thoughts!

Think the way I'm doing something is dumb and there's a better option? Is there something missing from an area I've deemed 'complete'? Let me know!

To make tracking these easier, please make separate threads for each piece of feedback.

**Comment by OmgImAlexis** [comment]
I'm really confused on how this is meant to work.

For example I'm trying to get an array of new posts from a sub but `client.subreddits.getNewPosts()` isn't a promise.

**Reply by thislooksfun** [reply]
`getNewPosts` (and most functions like it) returns a [Listing](https://thislooks.fun/snoots/docs/latest/classes/listing), not an array. This is a big change from how snoowrap works, so I get it's a bit confusing. There is an [example](https://thislooks.fun/snoots/docs/latest/index#examples) showing how to iterate through comments, but I'll add some more examples showing how it works.

The main thing is that if you want to iterate through a Listing, use [.each()](https://thislooks.fun/snoots/docs/latest/classes/listing#each).

**Reply by OmgImAlexis** [reply]
Why not just return an iterable?
Honestly I was expecting listing to extend iterable.

**Reply by thislooksfun** [reply]
Not extending Array was a very conscious choice as it massively reduces the boilerplate needed to use Listings and thus makes simple mistakes harder to make. Take the example of iterating through the all the replies to a post and printing them out. In snoowrap this is written as follows:

```ts
const post: Submission = await client.getSubmission("mpdc90").fetch();
for (const cmt of post.comments) {
  console.log(cmt.body);
}
```

Except that this will only print the first 50 comments. If you want more you have to remember to fetch them yourself:

```ts
const post: Submission = await client.getSubmission("mpdc90").fetch();
await post.comments.fetchAll();
for (const cmt of post.comments) {
  console.log(cmt.body);
}
```

But that doesn't work either, because `fetchAll()` is non-mutating. So you *actually* need to do this:

```ts
const post: Submission = await client.getSubmission("mpdc90").fetch();
const comments = await post.comments.fetchAll();
for (const cmt of comments) {
  console.log(cmt.body);
}
```

But what if you only want to interact with *some* of the comments, not all of them? After all, doing `fetchAll()` on a large post can result in *many* api calls (running the above code caused ***48*** more calls to the API just to fetch all the comments. Well in that case you'd need to fetch one page at a time:

```ts
const post: Submission = await client.getSubmission("m1kw8i").fetch();
let comments = post.comments;

while (!comments.isFinished) {
  for (const cmt of comments) {
    console.log(cmt.body);
  }

  comments = await comments.fetchMore({ amount: 50 });
}
```

Except that that doesn't work either! By default `fetchMore()` *appends to the existing array*, so this is processing each page multiple times. The *actual* code is as follows:

```ts
const post: Submission = await client.getSubmission("m1kw8i").fetch();
let comments = post.comments;

while (!comments.isFinished) {
  for (const cmt of comments) {
    console.log(cmt.body);
  }

  comments = await comments.fetchMore({ amount: 50, append: false });
}
```

That's a lot of boilerplate, and many details you have to remember to get it right. In snoots, this pattern is *way* simpler:

```ts
const post = await client.posts.fetch("m1kw8i");
await post.comments.each(c => console.log(c.body));
```

That's it. All the paging is handled automatically in the background, so you don't have to.

If you have a legitimate use case that can't be solved by using `.each()` (or `.eachPage()`), create a [new discussion](https://github.com/thislooksfun/snoots/discussions/new?category=Ideas) and we can discuss it further.

**Reply by OmgImAlexis** [reply]
“simple mistakes harder to make”.

You’re over complicating things for the point of making it easier. 

The fact that I have to learn a new way of handing the data just to use this library makes the barrier of entry higher than should be. After reading the comments in snoowrap that lead to this I had hoped this was going to be a more straightforward interface but it’s doing just the same, hiding so much that it becomes complicated to work with. Structures like iterables exist for a reason. I’m finding I’m needing to convert the result to one just to use this as I would any other library and I can see this exact issue driving others away.

I promise this isn’t an attack just an observation. I felt like I’ve spent more time trying to workout this then using the library.

**Reply by thislooksfun** [reply]
Again, if  you have a specific use case that the existing solution does not address, please [create a discussion](https://github.com/thislooksfun/snoots/discussions/new?category=Ideas) so we can address it! I personally have yet to run into a situation where the current solution doesn't make it easier, but if you're fighting it I'm open to changing the design.

My main reasoning is that Listings are *not* arrays, they are loaded piecemeal so treating them like arrays makes working with them harder imo.

**Reply by thislooksfun** [reply]
@OmgImAlexis What I'm trying (and apparently failing) to say is that Reddit's Listings are not actually arrays. The Reddit API, in its infinite wisdom, only gives back partial data for virtually *every* listing request. For example, here is the actual raw response when trying to get the newest posts from r/funny:<sup>1</sup>

```js
{
  kind: "Listing",
  data: {
    modhash: null,
    dist: 100,
    children: [
      {
        kind: "t3",
        data: { ... },
      },
      ... 99 more ...
    ],
    after: "t3_mpn4uh",
    before: null,
  },
}
```

As you can see though it only returns the first 100 items, and gives an `after` parameter to say "if you want more, fetch it again and say you want the items after t3_mpn4uh". So to get the full list of items in virtually *any* listing requires multiple API fetches. Because of this there are only 3 options I can see:

1. Fetch the entire Listing every time before handing it to the consumer. This is a bad idea because often people either don't care about the listing at all (think `comments` on Posts or the `replies` of Comments, if you only care about the body why waste effort fetching data that won't be used?) or only care about part of it (like looping until it finds something it's seen before).
2. Make the user fetch more when they need it. This is the pattern snoowrap uses, and I find it very clunky and error-prone, as described in a previous comment.
3. Make an opaque interface that cleanly iterates through the items without the user needing to worry about partial listings.

Making Listings into an iterable would necessitate one of the first two options, both of which are flawed and have more downsides than upsides. Option 3 (which is what snoots does) *does* make using array-based interactions impossible without an intermediate step, but I'm struggling to think of any real use cases that can't be covered by the existing api. What I meant by "if you have a specific use case" was "please share your code so we can see if there's anything I can do to help you clean it up", be that through a refactor on your end or a new api method on Listing.

My goal here *is* to make a more straightforward interface, and I strongly feel that hiding the messy (and quite frankly *ugly*) interactions needed to make Listings work is part of that.

---

<sup>1</sup> For the *full* response, you can see [here](https://gist.github.com/thislooksfun/cb6921ef6724421c761cdd89e36d7393), but be warned: it's very long.

**Reply by thislooksfun** [reply]
I wrote up a (hopefully) more detailed post over [here](https://github.com/thislooksfun/snoots/discussions/17). I'm going to lock this discussion as I never intended it harbor such in-depth discussions. If you want to continue this further (and I would encourage you to do so, since it seems you have legitimate feedback!) please [open a new discussion](https://github.com/thislooksfun/snoots/discussions/new?category=Ideas). That allows me to keep track of the conversation easier, and also mark when it has been resolved, which I can't do to a reply chain.

**Reply by thislooksfun** [reply]
@OmgImAlexis I'll admit it, I was wrong. I thought that iterators had to be synchronous, so it would be impossible to make work with the limitation at hand. I was completely unaware of `for await` and `Symbol.asyncIterator`. Without the existence of `for await` I think that my conclusions were valid, but since it exists I agree that it should be available. I'm in the process of implementing it and updating the docs now.

EDIT: Listings are async iterable as of [v1.0.0-dev.9](https://github.com/thislooksfun/snoots/releases/v1.0.0-dev.9)

**Comment by eli-rich** [comment]
When `fetch` becomes non-experimental, is there a plan to use that instead of got? Cuts out a dependency.

**Reply by thislooksfun** [reply]
[There is now](https://github.com/thislooksfun/snoots/issues/45)! I didn't actually know Node was adding a first-class `fetch()` implementation, so thanks for letting me know!
